### ISSUE TITLE
When setting subtitle id with --sid the option becomes ignored after switching subtitles

### ISSUE THREAD

**Issue by mlindner** [mpv-player/mpv#13670]
### Important Information

Provide following Information:
- mpv version = v0.37.0-280-gf746125769-dirty
- Linux Distribution and Version = Arch Linux
- Source of the mpv binary = self compiled
- GPU model, driver and version = not sure

### Reproduction steps

Use the following files: https://nyaa.si/view/1771619
Queue several files on the command line and pass --sid=3 (or any other value other than what's automatically selected).
Change to subtitle ID 4 with j, and then change back to subtitle ID 3 with J.
Go to the next file in the queue (with enter or >).
Try changing the subtitle, you'll find that somehow subtitle ID has been set to 2 instead of the 3 selected on the command line

### Expected behavior
Subtitle ID stays what is selected on the command line.

### Actual behavior
Subtitle ID gets reset to the automatically selected subtitle ID.

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/14544524/output.txt)

### Sample files

https://nyaa.si/view/1771619

**Comment by mlindner** [comment]
I'll note that I tried this with other files and I couldn't reproduce it, so this is probably file-specific. It might have something to do with some subtitle IDs having the 'forced' flag set in the mkv file.

**Comment by mlindner** [comment]
I tried setting simultaneously --subs-with-matching-audio=no --subs-match-os-language=no --subs-fallback=no --subs-fallback-forced=no and the issue still happens, but the behavior changes. As before, switching between files maintains the set --sid as long as you do not at any point manually change the subtitle track. If you do however, it again starts to ignore the command line again from that time point forward. While previously it would reset back to a default subtitle id that it chose for these files (sid=2) now it resets to no subtitle selected at all. So you still need to manually set the subtitle again after every file change in the playlist.

**Comment by hooke007** [comment]
https://mpv.io/manual/master/#options-reset-on-next-file

**Comment by Dudemanguy** [comment]
This is because `check_previous_track_selection` deliberately resets all track selection if the track layout is different in between files. Apparently this behavior always existed: https://github.com/mpv-player/mpv/commit/b19414f3bf0ae28461ef6d5048e9626e69d45319 but it does seem a little heavy handed.

**Comment by mlindner** [comment]
@Dudemanguy I was doing some debugging of my own while these comments were getting added, it seems like between release/0.28 and release/0.29 (and maybe again later) some code got reworked and the locations where mp_switch_track_n got called got reduced, and it bypassed it and it sets mpctx->opts->stream_id directly, (or in later versions calls mark_track_selection directly). Look at MP_PROPERTY_SWITCH in command.c in release/0.28 versus later versions.

**Comment by mlindner** [comment]
As a thought, it's probably still a good idea to reset the option by default when changing files, but I think the detection should be more flexible than just looking at a hash of the entire list of subtitle entries. It should instead hash just the individual subtitle entry selected and see if that matches after changing files. It's likely possible (as is the case here) that only the higher indexed subtitle entries are changing, but the lower indexed ones are the same.

**Comment by Dudemanguy** [comment]
Why so? I guess I'm just not understanding what the use case is supposed to be here but to me it seem utterly bizarre to wipe a user's set --sid just because they pressed `j`.

**Comment by mlindner** [comment]
> Why so? I guess I'm just not understanding what the use case is supposed to be here but to me it seem utterly bizarre to wipe a user's set --sid just because they pressed `j`.

To be clear, I'm still saying the behavior should be identical between if they pressed `j` or not. I'm just saying that we should keep/extend the behavior of resetting the subtitle selection if the selected track has fundamentally changed. If you had say an english language track on one file and that same subtitle id changes to a hungarian language track on the next file, that's probably not wanted, it should go back to auto selection.

**Comment by mlindner** [comment]
In other words, instead of a `track_layout_hash` function, do a `track_hash` function, and unify the behavior.

**Comment by Dudemanguy** [comment]
>If you had say an english language track on one file and that same subtitle id changes to a hungarian language track on the next file, that's probably not wanted, it should go back to auto selection.

I don't see the value in trying to make assumptions like this. We already have a gazillion subtitle options that are complicated enough as it is (e.g. one would simply use `--slang=en` for this usecase). Trying to guess if the user actually wants the sid that he previously changed to reset on the next file is not really worth it imo. Especially since you can already opt into this behavior with `--reset-on-next-file=sid`.

**Comment by mlindner** [comment]
> We already have a gazillion subtitle options that are complicated enough as it is (e.g. one would simply use `--slang=en` for this usecase).

Right, and it should go back to using that slang option (that's likely in their config file) after changing files.

**Comment by mlindner** [comment]
Actually thinking over this more, setting --sid is so strong that it should probably override everything. However cycling tracks with the keys should be "weaker" than the --sid option and get reset after changing files.

The comment is I think no longer completely accurate but it's a good refernce, but look at the comment above the `compare_track` function in loadfile.c. It's a really good ordered list of the previous thoughts on track priority.

**Comment by Dudemanguy** [comment]
I suppose the most gentle behavior change would be to rest to --sid (if it was set via commandline) rather than reset to -1 in this case. Would that work for you? Trying to distinguish between what is set via random scripts and other things sounds probably not worth it.

**Comment by pacexy** [comment]
As a temporary workaround, consider using [mpv-persist](https://github.com/pacexy/mpv-persist). This script:

1. Observes specific property changes.
2. Persists these properties in a directory-specific configuration file.
3. Loads them on the `file-loaded` event, which is necessary to handle unexpected property resets.
